### PR TITLE
Adds the hivebot invasion event

### PR DESCRIPTION
**Pull request by Cyberboss**
Follow up to #23398

:cl: Xhuis & Cyberboss
add: New hivebot invasion event
/:cl:

### CODE REVIEW

File: code/game/objects/items/devices/hivebot_crux.dm [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/obj/item/device/hivebot_crux
+	name = "hivebot crux"
+	desc = "A smooth, metallic metal cylinder with an indigo screen."
+	icon = 'icons/mob/hivebot.dmi'
+	icon_state = "hive_main-crash"
+	flags = CONDUCT | NOBLUDGEON
+	slot_flags = SLOT_BELT
+	materials = list(MAT_METAL = 500, MAT_GLASS = 1000)
+	origin_tech = "engineering=5;magnets=4;programming=4"
```
Comment (kevinz000) [inline code/game/objects/items/devices/hivebot_crux.dm]: shit levels tbh

**Comment by RemieRichards** [comment]
Hmmmmm I wonder why you're doing this :^)

**Comment by Cyberboss** [comment]
I literally just noticed the PR and thought it looked cool

**Comment by RemieRichards** [comment]
>I literally just noticed the PR and thought it looked cool

image

Sure :^)

**Comment by kevinz000** [comment]
lmao

File: code/modules/events/portal_storm.dm [hunk 2]
```diff
@@ -1,7 +1,7 @@
 /datum/round_event_control/portal_storm_syndicate
 	name = "Portal Storm: Syndicate Shocktroops"
 	typepath = /datum/round_event/portal_storm/syndicate_shocktroop
-	weight = 2
+	weight = 0
```
Comment (DoctoreMad) [inline code/modules/events/portal_storm.dm]: STELFY CHANGES HEHE

File: code/game/objects/items/devices/hivebot_crux.dm (hunk 1 above)
Comment (Xhuis) [inline code/game/objects/items/devices/hivebot_crux.dm]: The hivebot crux was originally intended to make a little helper companion that followed you around and would provide different stuff; i.e. if you were badly hurt, the little robot would detect that and stop any bleeding you might have and inject you with some very weak healing chemicals to tide you over. It's not intended for research only.

File: code/modules/events/portal_storm.dm (hunk 2 above)
Comment (Cyberboss) [inline code/modules/events/portal_storm.dm]: It's mainly a port of the original PR, I think @AnturK wanted this gone

Comment (GunHog) [inline code/modules/events/portal_storm.dm]: It needs to be in its own PR!

File: code/game/machinery/hivebot_swarm_core.dm [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	pixel_x = -32
+	pixel_y = -16
+	use_power = FALSE
+	obj_integrity = 250
+	max_integrity = 250
+	anchored = TRUE
+	density = TRUE
+	var/recharge_period = 0 //The number of ticks the core is waiting for and cannot do anything
+	var/mob/living/threat_to_swarm //The core's "target" for its attacks
+	var/hivebot_limit = 50 //Stop making hivebots if there are this many in a 7x7 square around us
```
Comment (octareenroon91) [inline code/game/machinery/hivebot_swarm_core.dm]: `range = 7` means 15x15. You had me worried here: 7x7 is 49 so a limit of 50 in that box would mean more than one hivebot per turf.

File: code/game/objects/items/devices/hivebot_crux.dm (hunk 1 above)
Comment (DoctoreMad) [inline code/game/objects/items/devices/hivebot_crux.dm]: Is that feature up?

File: code/modules/events/hivebot_invasion.dm [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/datum/round_event_control/hivebot_invasion //A hivebot swarm core impacts the station and warps in hivebots.
+	name = "Hivebot Invasion"
+	typepath = /datum/round_event/hivebot_invasion
+	weight = 2
```
Comment (DoctoreMad) [inline code/modules/events/hivebot_invasion.dm]: Isn't this weight lower than blob?

Comment (Chronitonity) [inline code/modules/events/hivebot_invasion.dm]: Yeah blobs weight is 5

File: code/game/machinery/hivebot_swarm_core.dm (hunk 3 above)
Comment (DoctoreMad) [inline code/game/machinery/hivebot_swarm_core.dm]: Besides shouldn't the limit be 40-30? Having 50 hivebots around the machine sounds a bit horrendous.

**Comment by Steelpoint** [comment]
This would be a interesting addition to see in game, its hard to conjecture how much of a impact it'll make until we see it in game, considering the nature of Hivebots they can be both very weak and somewhat strong at the same time depending on who is fighting them.

I think this has such a low weighting due to how disruptive it may be.

File: code/game/machinery/droneDispenser.dm [hunk 5]
```diff
@@ -129,6 +129,24 @@
...
+	begin_create_message = "closes, humming and whirring..."
+	end_create_message = "slams open, revealing a hivebot!"
+	recharge_sound = null
+	recharge_message = null
+
+/obj/machinery/droneDispenser/hivebot/invasion/ex_act(severity)
+	return //hivebot don give a fuck about bombs
+
+/obj/machinery/droneDispenser/hivebot/invasion/process()
+	dispense_type = pick(/mob/living/simple_animal/hostile/hivebot, /mob/living/simple_animal/hostile/hivebot/range, /mob/living/simple_animal/hostile/hivebot/rapid)
```
Comment (lzimann) [inline code/game/machinery/droneDispenser.dm]: make a new static list to pick from, instead of generating a new one every process

**Comment by Bawhoppen** [comment]
Needs to be less over the top

File: code/modules/events/portal_storm.dm (hunk 2 above)
Comment (nicbn) [inline code/modules/events/portal_storm.dm]: ~le stealth~

File: code/modules/events/hivebot_invasion.dm (hunk 4 above)
Comment (Steelpoint) [inline code/modules/events/hivebot_invasion.dm]: Weighting should be a little higher, considering Hivebots don't drop Syndi loot on death.

**Comment by Steelpoint** [comment]
I tried it out on local code and overall I think its a fair event.

When it triggers a massive alarm sounds warning the crew of a incoming object on collision course. After a few seconds the crew is informed of exactly where the core is going to crash into, so the people have plenty of time to evacuate the area to avoid getting crushed.

Once the core crashes it produces a weak Hivebot every two or so seconds unless someone is in range of the core, in which the core will attack a target instead of making Hivebots.

It is very disruptive but I think its fine as a uncommon/rare event.

**Comment by Steelpoint** [comment]
One thing I did notice however is that the Swarm Core will continue to constantly attack the same target over and over again, even if the target is dead or other targets present themselves.

**Comment by DoctoreMad** [comment]
I support this because there needs to be more events with huge warnings which are easy to avoid but instagib you if you're not paying attention
People need to look at announcements and they must learn it **BY FORCE**

**Comment by Bawhoppen** [comment]
We need LESS disruptive meme events, not more

**Comment by DoctoreMad** [comment]
Well
Some non-disruptive events with higher weight would be cool, yeah.

File: code/game/machinery/hivebot_swarm_core.dm [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	icon_state = "hivebot_swarm_core"
+	pixel_x = -32
+	pixel_y = -16
+	use_power = FALSE
+	obj_integrity = 250
+	max_integrity = 250
+	anchored = TRUE
+	density = TRUE
+	var/recharge_period = 0 //The number of ticks the core is waiting for and cannot do anything
+	var/mob/living/threat_to_swarm //The core's "target" for its attacks
```
Comment (ChangelingRain) [inline code/game/machinery/hivebot_swarm_core.dm]: what about mechs?

Comment (Xhuis) [inline code/game/machinery/hivebot_swarm_core.dm]: when I originally wrote this code, I excluded mechs under the fluff of the machine being too stupid to recognize them; that's also why it ignores borgs

Comment (DoctoreMad) [inline code/game/machinery/hivebot_swarm_core.dm]: Not very troublesome because borgs (usually) have no ranged lasers and if you have a DURAND/Gygax that thing is gonna go down anyways
That being said that makes us miss out on the chances for grand exosuit fights against the machine. Or the chance for a RIPLEY to just tear that shit apart.

Comment (Xhuis) [inline code/game/machinery/hivebot_swarm_core.dm]: The hivebots themselves are, but this big huge machine's sole purpose is to make more and to kill anything carbon-based that gets nearby. I like to headcanon that hivebots are 100% absorbed in efficiency and cut corners wherever possible, so it'd make sense that the hivebots themselves (who murder) can recognize threats, but the fabricator just recognizes the most obvious enemies.

**Comment by Steelpoint** [comment]
Bumping good content.

**Comment by Steelpoint** [comment]
Considering how rare this will be I think its fine as a event. The hivebots are only a massive threat if you try and confront the machine as they tend to cluster around the machine.

**Comment by Militaires** [comment]
>bumping

**Comment by KorPhaeron** [comment]
Ignoring mechs/borgs might be fun fluff but it will be boring gameplay

File: code/game/machinery/droneDispenser.dm [hunk 7]
```diff
@@ -129,6 +129,25 @@
...
+	end_create_message = "slams open, revealing a hivebot!"
+	recharge_sound = null
+	recharge_message = null
+
+/obj/machinery/droneDispenser/hivebot/invasion/ex_act(severity)
+	return //hivebot don give a fuck about bombs
+
+/obj/machinery/droneDispenser/hivebot/invasion/process()
+	var/static/list/hivebot_dispensable_types = list(/mob/living/simple_animal/hostile/hivebot, /mob/living/simple_animal/hostile/hivebot/range, /mob/living/simple_animal/hostile/hivebot/rapid)
+	dispense_type = pick(hivebot_dispensable_types)
```
Comment (AnturK) [inline code/game/machinery/droneDispenser.dm]: Shouldn't this be done at creation instead of every process if it's random anyway.

File: code/game/machinery/hivebot_swarm_core.dm [hunk 8]
```diff
@@ -0,0 +1,111 @@
+/obj/machinery/hivebot_swarm_core //A hulking machine armed with several weapons. Creates hivebots and defends itself with a buzzsaw and mounted weapons.
+	name = "hivebot swarm core"
+	desc = "An incredibly dangerous machine with random weapons haphazardly attached with spot welds. Its main body seems impervious to damage, but the head is vulnerable."
+	icon = 'icons/effects/96x96.dmi'
+	icon_state = "hivebot_swarm_core"
```
Comment (AnturK) [inline code/game/machinery/hivebot_swarm_core.dm]: Post pictures please.

File: code/game/machinery/hivebot_swarm_core.dm [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	new/obj/structure/fluff/hivebot_swarm_core(get_turf(src))
+	return ..()
+
+/obj/machinery/hivebot_swarm_core/examine(mob/user)
+	..()
+	if(recharge_period)
+		to_chat(user, "<span class='warning'>It seems to be recharging.</span>")
+
+/obj/machinery/hivebot_swarm_core/process()
+	if(recharge_period)
```
Comment (AnturK) [inline code/game/machinery/hivebot_swarm_core.dm]: Could just make this timer instead of process

File: code/modules/events/hivebot_invasion.dm [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	earliest_start = 18000
+
+/datum/round_event/hivebot_invasion
+	announceWhen	= 0
+	startWhen		= 20
+	var/obj/effect/landmark/invasion_point
+
+/datum/round_event/hivebot_invasion/setup()
+	invasion_point = pick(GLOB.generic_event_spawns)
+	message_admins("Hivebot invasion point set to [get_area(invasion_point)]! ([invasion_point.x],[invasion_point.y],[invasion_point.z] - <a href='?_src_=holder;adminplayerobservecoodjump=1;X=[invasion_point.x];Y=[invasion_point.y];Z=[invasion_point.z]'>JMP</a>)")
```
Comment (AnturK) [inline code/modules/events/hivebot_invasion.dm]: Pretty sure there's admin turf jump href macro

File: code/modules/mob/living/simple_animal/hostile/hivebot.dm [hunk 11]
```diff
@@ -52,6 +52,30 @@
...
+	icon_state = "EngBot"
+	health = 25
+	maxHealth = 25
+	melee_damage_lower = 0
+	melee_damage_upper = 0
+	ranged = 1
+	retreat_distance = 2
+	minimum_distance = 2
+
+/mob/living/simple_animal/hostile/hivebot/engineer/Life()
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/hostile/hivebot.dm]: This should be in handle_automated_action not in Life

File: code/game/machinery/hivebot_swarm_core.dm (hunk 8 above)
Comment (PraiseRatvar) [inline code/game/machinery/hivebot_swarm_core.dm]: http://i.imgur.com/oZ3J1M4.png

**Comment by DoctoreMad** [comment]
@KorPhaeron Hivebots will still attack borgs and exosuits. Actually, I think exosuits enrage mobs on-sight because they have a higher attack priority or something?

File: code/game/machinery/hivebot_swarm_core.dm (hunk 9 above)
Comment (Cyberboss) [inline code/game/machinery/hivebot_swarm_core.dm]: It only dispenses if there's something nearby, that's why it processes along with the cooldown

File: code/modules/events/hivebot_invasion.dm (hunk 10 above)
Comment (lzimann) [inline code/modules/events/hivebot_invasion.dm]: `ADMIN_COORDJMP(turf)`

**Comment by Armhulen** [comment]
Stale!

**Comment by Cyberboss** [comment]
Not stale, I just have to merge the fucking dmis

**Comment by Steelpoint** [comment]
Bump. Good PR needs fixing.

**Comment by lzimann** [comment]
@Cyberboss

**Comment by Cyberboss** [comment]
Don't close, I'll try this shit now

File: code/game/machinery/hivebot_swarm_core.dm [hunk 12]
```diff
@@ -0,0 +1,124 @@
+/obj/machinery/hivebot_swarm_core //A hulking machine armed with several weapons. Creates hivebots and defends itself with a buzzsaw and mounted weapons.
+	name = "hivebot swarm core"
+	desc = "An incredibly dangerous machine with random weapons haphazardly attached with spot welds. Its main body seems impervious to damage, but the head is vulnerable."
+	icon = 'icons/effects/96x96.dmi'
+	icon_state = "hivebot_swarm_core"
+	pixel_x = -32
+	pixel_y = -16
+	use_power = FALSE
+	obj_integrity = 250
```
Comment (optimumtact) [inline code/game/machinery/hivebot_swarm_core.dm]: remove this line

File: code/game/machinery/hivebot_swarm_core.dm [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	for(var/mob/living/silicon/S in view(7, src))
+		potential_targets += S
+	for(var/obj/mecha/M in view(7, src))
+		potential_targets += M
+	for(var/mob/living/L in potential_targets)
+		if(L.stat)
+			potential_targets -= L
+		else
+			for(var/faction in L.faction)
+				if(faction == "hivebot")
```
Comment (optimumtact) [inline code/game/machinery/hivebot_swarm_core.dm]: ? you should just use the faction var, it's an atom var, so you can set it on the machine iirc

then just do if(src.faction in L.faction)

Comment (optimumtact) [inline code/game/machinery/hivebot_swarm_core.dm]: otherwise just do if("hivebot" in L.faction)

**Comment by Armhulen** [comment]
almost at a month guys! derelict pr ftw!!!!

**Comment by Cyberboss** [comment]
Was this tested?

**Comment by Xhuis** [comment]
Considering my testing shows that hivebot cores can buzzsaw you from several rooms away, it doesn't seem to have been!

**Comment by optimumtact** [comment]
I guess I Just assumed either of you two had you know, tested the code before spending all this time developing it?

**Comment by optimumtact** [comment]
er after, not before

**Comment by Xhuis** [comment]
I left it off in a vastly incomplete state. I never had a chance to properly test it before I lost interest. Cyber picked it up, but I'm not sure how he worked with it since then.
